tensorflow/python/ops/bincount_ops_test.py

```diff
@@ -163,6 +165,188 @@ def test_dense_input(self,
+    def f(x,
+          weights=weights,
+          minlength=minlength,
+          maxlength=maxlength,
+          binary_output=binary_output,
+          axis=axis):
+      y = bincount_ops.bincount(
+            x,
+            weights=weights,
+            minlength=minlength,
+            maxlength=maxlength,
+            binary_output=binary_output,
+            axis=axis
+        )
+      return y
+    y = f(x,           
+          weights=weights,
+          minlength=minlength,
+          maxlength=maxlength,
+          binary_output=binary_output,
+          axis=axis)
+    self.assertAllEqual(expected_values, y)
+  @parameterized.named_parameters(    
+  {
+      "testcase_name": "_no_maxlength_small",
+      "x": np.random.randint(200, size=(200, 200), dtype=np.int32)
+  }, {
+      "testcase_name": "_no_maxlength_medium",
+      "x": np.random.randint(500, size=(500, 500), dtype=np.int32)
+  }, {
+      "testcase_name": "_no_maxlength_large",
+      "x": np.random.randint(100, size=(1000, 1000), dtype=np.int32)
+  })
+  @test_util.disable_mlir_bridge('TODO: ?')
+  def test_compiled_dense_perf(self,
+                       x,
+                       minlength=None,
+                       maxlength=None,
+                       binary_output=False,
+                       weights=None,
+                       axis=-1):
+
+    @def_function.function(jit_compile=True)
+    def f_compiled(x,
+          weights=weights,
+          minlength=minlength,
+          maxlength=maxlength,
+          binary_output=binary_output,
+          axis=axis):
+      y = bincount_ops.bincount(
+            x,
+            weights=weights,
+            minlength=minlength,
+            maxlength=maxlength,
+            binary_output=binary_output,
+            axis=axis
+        )
+      return y
+
+    @def_function.function(jit_compile=False)
+    def f(x,
+          weights=weights,
+          minlength=minlength,
+          maxlength=maxlength,
+          binary_output=binary_output,
+          axis=axis):
+      y = bincount_ops.bincount(
+            x,
+            weights=weights,
+            minlength=minlength,
+            maxlength=maxlength,
+            binary_output=binary_output,
+            axis=axis
+        )
+      return y
+
+    lambda_f = lambda: f(x,           
+                weights=weights,
+                minlength=minlength,
+                maxlength=maxlength,
+                binary_output=binary_output,
+                axis=axis)
+
+    lambda_fc = lambda: f_compiled(x,           
+                        weights=weights,
+                        minlength=minlength,
+                        maxlength=maxlength,
+                        binary_output=binary_output,
+                        axis=axis)
+    # warm-up
+    lambda_f(); lambda_fc()
+    not_compiled = timeit.timeit(lambda_f, number=10)
+    compiled = timeit.timeit(lambda_fc, number=10)
+    print("XLA JIT -> compiled: %f | not compiled: %f" % 
```

Comment:
- **P100 GPU**:
XLA JIT -> compiled: 0.068937 | not compiled: 0.036515
XLA JIT -> compiled: 0.024318 | not compiled: 0.016768
XLA JIT -> compiled: 0.008323 | not compiled: 0.007022

- **CPU Intel(R) Xeon(R) CPU @ 2.30GHz   (Intel Haswell)**:
XLA JIT -> compiled: 0.068736 | not compiled: 0.037503
XLA JIT -> compiled: 0.021457 | not compiled: 0.016532
XLA JIT -> compiled: 0.008780 | not compiled: 0.007548

Reply:
GPU is going in timeout out locally and on the CI:

```MLIR
HloModule a_inference_f_compiled_865__.75

%bincount_Max-reduction.5 (x.6: s32[], y.7: s32[]) -> s32[] {
  %x.6 = s32[] parameter(0)
  %y.7 = s32[] parameter(1)
  ROOT %maximum.8 = s32[] maximum(s32[] %x.6, s32[] %y.7)
}

%update_rank2.23 (param.24: (s64[], s64[], s64[], s32[1000,100], s32[], /*index=5*/s32[0])) -> (s32[], s32[1000,100]) {
  %param.24 = (s64[], s64[], s64[], s32[1000,100]{1,0}, s32[], /*index=5*/s32[0]{0}) parameter(0)
  %get-tuple-element.27 = s64[] get-tuple-element((s64[], s64[], s64[], s32[1000,100]{1,0}, s32[], /*index=5*/s32[0]{0}) %param.24), index=2
  %get-tuple-element.30 = s32[0]{0} get-tuple-element((s64[], s64[], s64[], s32[1000,100]{1,0}, s32[], /*index=5*/s32[0]{0}) %param.24), index=5
  %get-tuple-element.29 = s32[] get-tuple-element((s64[], s64[], s64[], s32[1000,100]{1,0}, s32[], /*index=5*/s32[0]{0}) %param.24), index=4
  %constant.31 = s32[] constant(1)
  %add.32 = s32[] add(s32[] %get-tuple-element.29, s32[] %constant.31)
  %get-tuple-element.28 = s32[1000,100]{1,0} get-tuple-element((s64[], s64[], s64[], s32[1000,100]{1,0}, s32[], /*index=5*/s32[0]{0}) %param.24), index=3
  %reshape.33 = s32[1,1]{1,0} reshape(s32[] %add.32)
  %get-tuple-element.26 = s64[] get-tuple-element((s64[], s64[], s64[], s32[1000,100]{1,0}, s32[], /*index=5*/s32[0]{0}) %param.24), index=1
  %get-tuple-element.25 = s64[] get-tuple-element((s64[], s64[], s64[], s32[1000,100]{1,0}, s32[], /*index=5*/s32[0]{0}) %param.24), index=0
  %dynamic-update-slice.34 = s32[1000,100]{1,0} dynamic-update-slice(s32[1000,100]{1,0} %get-tuple-element.28, s32[1,1]{1,0} %reshape.33, s64[] %get-tuple-element.26, s64[] %get-tuple-element.25)
  ROOT %tuple.35 = (s32[], s32[1000,100]{1,0}) tuple(s32[] %add.32, s32[1000,100]{1,0} %dynamic-update-slice.34)
}

%no_update_rank2.36 (param.37: (s64[], s64[], s64[], s32[1000,100], s32[], /*index=5*/s32[0])) -> (s32[], s32[1000,100]) {
  %param.37 = (s64[], s64[], s64[], s32[1000,100]{1,0}, s32[], /*index=5*/s32[0]{0}) parameter(0)
  %get-tuple-element.39 = s32[] get-tuple-element((s64[], s64[], s64[], s32[1000,100]{1,0}, s32[], /*index=5*/s32[0]{0}) %param.37), index=4
  %get-tuple-element.38 = s32[1000,100]{1,0} get-tuple-element((s64[], s64[], s64[], s32[1000,100]{1,0}, s32[], /*index=5*/s32[0]{0}) %param.37), index=3
  ROOT %tuple.40 = (s32[], s32[1000,100]{1,0}) tuple(s32[] %get-tuple-element.39, s32[1000,100]{1,0} %get-tuple-element.38)
}

%body.41 (param.42: (s64[], s32[1000,1000], s32[1000,100], s32[0])) -> (s64[], s32[1000,1000], s32[1000,100], s32[0]) {
  %param.42 = (s64[], s32[1000,1000]{1,0}, s32[1000,100]{1,0}, s32[0]{0}) parameter(0)
  %get-tuple-element.44 = s32[1000,1000]{1,0} get-tuple-element((s64[], s32[1000,1000]{1,0}, s32[1000,100]{1,0}, s32[0]{0}) %param.42), index=1
  %get-tuple-element.43 = s64[] get-tuple-element((s64[], s32[1000,1000]{1,0}, s32[1000,100]{1,0}, s32[0]{0}) %param.42), index=0
  %constant.47 = s64[] constant(1000)
  %divide.48 = s64[] divide(s64[] %get-tuple-element.43, s64[] %constant.47)
  %remainder.49 = s64[] remainder(s64[] %get-tuple-element.43, s64[] %constant.47)
  %dynamic-slice.50 = s32[1,1]{1,0} dynamic-slice(s32[1000,1000]{1,0} %get-tuple-element.44, s64[] %divide.48, s64[] %remainder.49), dynamic_slice_sizes={1,1}
  %reshape.51 = s32[] reshape(s32[1,1]{1,0} %dynamic-slice.50)
  %convert.52 = s64[] convert(s32[] %reshape.51)
  %convert.56 = s64[] convert(s64[] %convert.52)
  %constant.55 = s64[] constant(100)
  %compare.57 = pred[] compare(s64[] %convert.56, s64[] %constant.55), direction=LT
  %get-tuple-element.45 = s32[1000,100]{1,0} get-tuple-element((s64[], s32[1000,1000]{1,0}, s32[1000,100]{1,0}, s32[0]{0}) %param.42), index=2
  %dynamic-slice.53 = s32[1,1]{1,0} dynamic-slice(s32[1000,100]{1,0} %get-tuple-element.45, s64[] %divide.48, s64[] %convert.52), dynamic_slice_sizes={1,1}
  %reshape.54 = s32[] reshape(s32[1,1]{1,0} %dynamic-slice.53)
  %get-tuple-element.46 = s32[0]{0} get-tuple-element((s64[], s32[1000,1000]{1,0}, s32[1000,100]{1,0}, s32[0]{0}) %param.42), index=3
  %tuple.58 = (s64[], s64[], s64[], s32[1000,100]{1,0}, s32[], /*index=5*/s32[0]{0}) tuple(s64[] %convert.56, s64[] %divide.48, s64[] %remainder.49, s32[1000,100]{1,0} %get-tuple-element.45, s32[] %reshape.54, /*index=5*/s32[0]{0} %get-tuple-element.46)
  %conditional.59 = (s32[], s32[1000,100]{1,0}) conditional(pred[] %compare.57, (s64[], s64[], s64[], s32[1000,100]{1,0}, s32[], /*index=5*/s32[0]{0}) %tuple.58, (s64[], s64[], s64[], s32[1000,100]{1,0}, s32[], /*index=5*/s32[0]{0}) %tuple.58), true_computation=%update_rank2.23, false_computation=%no_update_rank2.36
  %get-tuple-element.60 = s32[] get-tuple-element((s32[], s32[1000,100]{1,0}) %conditional.59), index=0
  %constant.62 = s64[] constant(1)
  %add.63 = s64[] add(s64[] %get-tuple-element.43, s64[] %constant.62)
  %get-tuple-element.61 = s32[1000,100]{1,0} get-tuple-element((s32[], s32[1000,100]{1,0}) %conditional.59), index=1
  ROOT %tuple.64 = (s64[], s32[1000,1000]{1,0}, s32[1000,100]{1,0}, s32[0]{0}) tuple(s64[] %add.63, s32[1000,1000]{1,0} %get-tuple-element.44, s32[1000,100]{1,0} %get-tuple-element.61, s32[0]{0} %get-tuple-element.46)
}

%condition.65 (param.66: (s64[], s32[1000,1000], s32[1000,100], s32[0])) -> pred[] {
  %constant.68 = s64[] constant(1000000)
  %param.66 = (s64[], s32[1000,1000]{1,0}, s32[1000,100]{1,0}, s32[0]{0}) parameter(0)
  %get-tuple-element.67 = s64[] get-tuple-element((s64[], s32[1000,1000]{1,0}, s32[1000,100]{1,0}, s32[0]{0}) %param.66), index=0
  ROOT %compare.69 = pred[] compare(s64[] %constant.68, s64[] %get-tuple-element.67), direction=GT
}

ENTRY %a_inference_f_compiled_865__.75 () -> s32[1000,100] {
  %constant.11 = s32[] constant(1), metadata={op_type="AddV2" op_name="bincount/add" source_file="/root/.cache/bazel/_bazel_root/2c92b5569ddded7b3a6bd5e139451b60/execroot/org_tensorflow/bazel-out/k8-opt/bin/tensorflow/python/bincount_ops_test_gpu.runfiles/org_tensorflow/tensorflow/python/ops/bincount_ops_test.py" source_line=302}
  %constant.14 = s32[] constant(1), metadata={op_type="Mul" op_name="bincount/mul" source_file="/root/.cache/bazel/_bazel_root/2c92b5569ddded7b3a6bd5e139451b60/execroot/org_tensorflow/bazel-out/k8-opt/bin/tensorflow/python/bincount_ops_test_gpu.runfiles/org_tensorflow/tensorflow/python/ops/bincount_ops_test.py" source_line=302}
  %constant.15 = s32[] constant(1), metadata={op_type="Mul" op_name="bincount/mul" source_file="/root/.cache/bazel/_bazel_root/2c92b5569ddded7b3a6bd5e139451b60/execroot/org_tensorflow/bazel-out/k8-opt/bin/tensorflow/python/bincount_ops_test_gpu.runfiles/org_tensorflow/tensorflow/python/ops/bincount_ops_test.py" source_line=302}
  %constant.1 = s32[1000,1000]{1,0} constant({...}), metadata={op_type="Max" op_name="bincount/Max" source_file="/root/.cache/bazel/_bazel_root/2c92b5569ddded7b3a6bd5e139451b60/execroot/org_tensorflow/bazel-out/k8-opt/bin/tensorflow/python/bincount_ops_test_gpu.runfiles/org_tensorflow/tensorflow/python/ops/bincount_ops_test.py" source_line=302}
  %convert.2 = s32[1000,1000]{1,0} convert(s32[1000,1000]{1,0} %constant.1), metadata={op_type="Max" op_name="bincount/Max" source_file="/root/.cache/bazel/_bazel_root/2c92b5569ddded7b3a6bd5e139451b60/execroot/org_tensorflow/bazel-out/k8-opt/bin/tensorflow/python/bincount_ops_test_gpu.runfiles/org_tensorflow/tensorflow/python/ops/bincount_ops_test.py" source_line=302}
  %constant.3 = s32[] constant(-2147483648), metadata={op_type="Max" op_name="bincount/Max" source_file="/root/.cache/bazel/_bazel_root/2c92b5569ddded7b3a6bd5e139451b60/execroot/org_tensorflow/bazel-out/k8-opt/bin/tensorflow/python/bincount_ops_test_gpu.runfiles/org_tensorflow/tensorflow/python/ops/bincount_ops_test.py" source_line=302}
  %convert.4 = s32[] convert(s32[] %constant.3), metadata={op_type="Max" op_name="bincount/Max" source_file="/root/.cache/bazel/_bazel_root/2c92b5569ddded7b3a6bd5e139451b60/execroot/org_tensorflow/bazel-out/k8-opt/bin/tensorflow/python/bincount_ops_test_gpu.runfiles/org_tensorflow/tensorflow/python/ops/bincount_ops_test.py" source_line=302}
  %reduce.9 = s32[] reduce(s32[1000,1000]{1,0} %convert.2, s32[] %convert.4), dimensions={0,1}, to_apply=%bincount_Max-reduction.5, metadata={op_type="Max" op_name="bincount/Max" source_file="/root/.cache/bazel/_bazel_root/2c92b5569ddded7b3a6bd5e139451b60/execroot/org_tensorflow/bazel-out/k8-opt/bin/tensorflow/python/bincount_ops_test_gpu.runfiles/org_tensorflow/tensorflow/python/ops/bincount_ops_test.py" source_line=302}
  %convert.10 = s32[] convert(s32[] %reduce.9), metadata={op_type="Max" op_name="bincount/Max" source_file="/root/.cache/bazel/_bazel_root/2c92b5569ddded7b3a6bd5e139451b60/execroot/org_tensorflow/bazel-out/k8-opt/bin/tensorflow/python/bincount_ops_test_gpu.runfiles/org_tensorflow/tensorflow/python/ops/bincount_ops_test.py" source_line=302}
  %constant.12 = s32[] constant(1), metadata={op_type="AddV2" op_name="bincount/add" source_file="/root/.cache/bazel/_bazel_root/2c92b5569ddded7b3a6bd5e139451b60/execroot/org_tensorflow/bazel-out/k8-opt/bin/tensorflow/python/bincount_ops_test_gpu.runfiles/org_tensorflow/tensorflow/python/ops/bincount_ops_test.py" source_line=302}
  %add.13 = s32[] add(s32[] %convert.10, s32[] %constant.12), metadata={op_type="AddV2" op_name="bincount/add" source_file="/root/.cache/bazel/_bazel_root/2c92b5569ddded7b3a6bd5e139451b60/execroot/org_tensorflow/bazel-out/k8-opt/bin/tensorflow/python/bincount_ops_test_gpu.runfiles/org_tensorflow/tensorflow/python/ops/bincount_ops_test.py" source_line=302}
  %multiply.16 = s32[] multiply(s32[] %constant.15, s32[] %add.13), metadata={op_type="Mul" op_name="bincount/mul" source_file="/root/.cache/bazel/_bazel_root/2c92b5569ddded7b3a6bd5e139451b60/execroot/org_tensorflow/bazel-out/k8-opt/bin/tensorflow/python/bincount_ops_test_gpu.runfiles/org_tensorflow/tensorflow/python/ops/bincount_ops_test.py" source_line=302}
  %constant.19 = s64[] constant(0), metadata={op_type="DenseBincount" op_name="bincount/DenseBincount" source_file="/root/.cache/bazel/_bazel_root/2c92b5569ddded7b3a6bd5e139451b60/execroot/org_tensorflow/bazel-out/k8-opt/bin/tensorflow/python/bincount_ops_test_gpu.runfiles/org_tensorflow/tensorflow/python/ops/bincount_ops_test.py" source_line=302}
  %constant.17 = s32[1000,1000]{1,0} constant({...}), metadata={op_type="DenseBincount" op_name="bincount/DenseBincount" source_file="/root/.cache/bazel/_bazel_root/2c92b5569ddded7b3a6bd5e139451b60/execroot/org_tensorflow/bazel-out/k8-opt/bin/tensorflow/python/bincount_ops_test_gpu.runfiles/org_tensorflow/tensorflow/python/ops/bincount_ops_test.py" source_line=302}
  %constant.20 = s32[] constant(0), metadata={op_type="DenseBincount" op_name="bincount/DenseBincount" source_file="/root/.cache/bazel/_bazel_root/2c92b5569ddded7b3a6bd5e139451b60/execroot/org_tensorflow/bazel-out/k8-opt/bin/tensorflow/python/bincount_ops_test_gpu.runfiles/org_tensorflow/tensorflow/python/ops/bincount_ops_test.py" source_line=302}
  %broadcast.21 = s32[1000,100]{1,0} broadcast(s32[] %constant.20), dimensions={}, metadata={op_type="DenseBincount" op_name="bincount/DenseBincount" source_file="/root/.cache/bazel/_bazel_root/2c92b5569ddded7b3a6bd5e139451b60/execroot/org_tensorflow/bazel-out/k8-opt/bin/tensorflow/python/bincount_ops_test_gpu.runfiles/org_tensorflow/tensorflow/python/ops/bincount_ops_test.py" source_line=302}
  %constant.18 = s32[0]{0} constant({}), metadata={op_type="DenseBincount" op_name="bincount/DenseBincount" source_file="/root/.cache/bazel/_bazel_root/2c92b5569ddded7b3a6bd5e139451b60/execroot/org_tensorflow/bazel-out/k8-opt/bin/tensorflow/python/bincount_ops_test_gpu.runfiles/org_tensorflow/tensorflow/python/ops/bincount_ops_test.py" source_line=302}
  %tuple.22 = (s64[], s32[1000,1000]{1,0}, s32[1000,100]{1,0}, s32[0]{0}) tuple(s64[] %constant.19, s32[1000,1000]{1,0} %constant.17, s32[1000,100]{1,0} %broadcast.21, s32[0]{0} %constant.18), metadata={op_type="DenseBincount" op_name="bincount/DenseBincount" source_file="/root/.cache/bazel/_bazel_root/2c92b5569ddded7b3a6bd5e139451b60/execroot/org_tensorflow/bazel-out/k8-opt/bin/tensorflow/python/bincount_ops_test_gpu.runfiles/org_tensorflow/tensorflow/python/ops/bincount_ops_test.py" source_line=302}
  %while.70 = (s64[], s32[1000,1000]{1,0}, s32[1000,100]{1,0}, s32[0]{0}) while((s64[], s32[1000,1000]{1,0}, s32[1000,100]{1,0}, s32[0]{0}) %tuple.22), condition=%condition.65, body=%body.41, metadata={op_type="DenseBincount" op_name="bincount/DenseBincount" source_file="/root/.cache/bazel/_bazel_root/2c92b5569ddded7b3a6bd5e139451b60/execroot/org_tensorflow/bazel-out/k8-opt/bin/tensorflow/python/bincount_ops_test_gpu.runfiles/org_tensorflow/tensorflow/python/ops/bincount_ops_test.py" source_line=302}
  %get-tuple-element.71 = s32[1000,100]{1,0} get-tuple-element((s64[], s32[1000,1000]{1,0}, s32[1000,100]{1,0}, s32[0]{0}) %while.70), index=2, metadata={op_type="DenseBincount" op_name="bincount/DenseBincount" source_file="/root/.cache/bazel/_bazel_root/2c92b5569ddded7b3a6bd5e139451b60/execroot/org_tensorflow/bazel-out/k8-opt/bin/tensorflow/python/bincount_ops_test_gpu.runfiles/org_tensorflow/tensorflow/python/ops/bincount_ops_test.py" source_line=302}
  %reshape.72 = s32[1000,100]{1,0} reshape(s32[1000,100]{1,0} %get-tuple-element.71), metadata={op_name="XLA_Retvals"}
  %tuple.73 = (s32[1000,100]{1,0}) tuple(s32[1000,100]{1,0} %reshape.72), metadata={op_name="XLA_Retvals"}
  ROOT %get-tuple-element.74 = s32[1000,100]{1,0} get-tuple-element((s32[1000,100]{1,0}) %tuple.73), index=0, metadata={op_name="XLA_Retvals"}
}
```